qupath-extension-bioformats/src/main/java/qupath/lib/images/servers/bioformats/readers/ReaderWrapper.java

```diff
@@ -0,0 +1,39 @@
+package qupath.lib.images.servers.bioformats.readers;
+
+import qupath.lib.images.servers.TileRequest;
+
+import java.awt.image.*;
+import java.io.IOException;
+
+/**
+ * <p>
+ *     Wrapper around an image reader. It should be used with a {@link ReaderPool}.
+ * </p>
+ * <p>A reader must be {@link #close() closed} once no longer used.</p>
+ */
+public interface ReaderWrapper extends AutoCloseable {
+
+    /**
+     * Reads a tile from the image.
+     *
+     * @param tileRequest  the parameters defining the tile
+     * @param series  some images contain multiple image stacks or experiments within one file.
+     *                The one to use is defined by this parameter
+     * @param numberOfChannels  the number of channels of this image
+     * @param colorModel  the color model to use with this image
+     * @return the image corresponding to these parameters
+     * @throws IOException when a reading error occurs
+     */
+    BufferedImage getImage(TileRequest tileRequest, int series, int numberOfChannels, boolean isRGB, ColorModel colorModel) throws IOException;
```

Comment:
Do you remember the logic behind the `OMEReaderWrapper.createSampleModel()` function? Supporting arbitrary combinations of channels will require to edit this function, and I'm a bit confused with this part:
```java
int[] offsets = new int[numberOfChannels];
int[] bandInds = new int[numberOfChannels];
int ind = 0;

int channelCount = getChannelCount(series);
for (int cInd = 0; cInd < channelCount; cInd++) {
    int nSamples = getChannelSamplesPerPixel(series, cInd);
    for (int s = 0; s < nSamples; s++) {
        bandInds[ind] = cInd;
        if (isInterleaved()) {
            offsets[ind] = s;
        } else {
            offsets[ind] = s * tileRequest.getTileWidth() * tileRequest.getTileHeight();
        }
        ind++;
    }
}
// TODO: Check this! It works for the only test image I have... (2 channels with 3 samples each)
// I would guess it fails if pixelStride does not equal nSamples, and if nSamples is different for different 'channels' -
// but I don't know if this occurs in practice.
// If it does, I don't see a way to use a ComponentSampleModel... which could complicate things quite a bit
int pixelStride = numberOfChannels / effectiveC;
int scanlineStride = pixelStride * tileRequest.getTileWidth();
return new ComponentSampleModel(
        dataBuffer.getDataType(),
        tileRequest.getTileWidth(),
        tileRequest.getTileHeight(),
        pixelStride,
        scanlineStride,
        bandInds,
        offsets
);
```
In particular, what is the difference between `numberOfChannels`, `effectiveC`, and `channelCount`?

Reply:
I struggle to remember the logic exactly - only that it was really difficult to get it right... 

The general idea is that there are three confusingly inter-related terms that I will attempt to define (at least according to my best understanding):

* `Channels` are essentially [as described here](https://bioimagebook.github.io/chapters/1-concepts/4-colors/colors.html); different 2D images that are usually viewed using different colors that are mixed together (e.g. red, green & blue)
* `Bands` are arrays of pixels, which *can* correspond to channels - i.e. each channel has a band - but don't *have* to, because
* `Samples` provide a way to store multiple channels in a single band

In the end, if you store your channels in a single band then the values are interleaved (e.g. `RGBRGBRGB...` - this would be 1 band with 3 samples per pixel).

Alternatively, you might store your channels in separate bands, which can be referred to as *planar* storage (`RRRR.... GGGG.... BBB...` - 3 bands with 1 sample per pixel).

So to construct a `BufferedImage`, you need a `SampleModel` in order to be able to make sense of what is actually contained within the bands.

As far as I recall, the `scanlineStride` is generally the width of the image - *but might not be*. The main time when it isn't is if you have a [subimage](https://docs.oracle.com/en/java/javase/20/docs/api/java.desktop/java/awt/image/BufferedImage.html#getSubimage(int,int,int,int)) that is backed by the same data as the original, larger image.

QuPath refers to `channels` as the 'intuitive' concept, but the actual technical storage relies upon `bands` and `samples`.

You don't necessarily have to bother with this now - especially if there is a chance of regression, since it seems to be working ok. If the API supports arbitrary channels, I think it's ok to throw an `UnsupportedOperationException` if it's actually called (and document this) - since it can just be a placeholder for the future. QuPath itself doesn't have an API to request fewer channels in `ImageServer` anyway, but it will likely need one in the future because it's inefficient to request 50 channels all at once if they aren't needed.

If you do change this logic at all, one way to test it is working is to export multichannel images using [code like this](https://forum.image.sc/t/exporting-selected-planes-with-qupath/77610/5), with `channelsInterleaved()` turned on and off.

Reply:
Just to add - you'll see `bands` and `samples` in the Java documentation, but I think Bio-Formats tends to refer to `channels` and `samples`. Unfortunately, in this case `channels` doesn't necessarily match which QuPath's idea of channels.

I *think* you should probably consider a Bio-Formats `channel` to be like a `BufferedImage` `band`.

Reply:
OK, thanks for the explanation.

If that's fine with you, I can change only the public API for now (and throw `UnsupportedOperationException` like you mentioned) and will come back to this later (like when a first version of the omero extension will be ready)